src/realm/alloc_slab.cpp

```diff
@@ -1151,6 +1157,113 @@ void SlabAlloc::set_file_format_version(int file_format_version) noexcept
+
+    size_t fsiz = get_file().get_size();
+    size_t head_size = fsiz >= block_size ? block_size : block_size - fsiz;
+    size_t tail_size = fsiz >= 2 * block_size ? block_size : (fsiz > head_size ? fsiz - head_size : 0);
+    size_t aligned_tail_offset = round_down(fsiz - tail_size, page_size());
+    char* map;
+
+    try {
+        // map() can fail if non-Realm process shortens the file size between get_size() and map()
+        map = (char*)get_file().map(File::access_ReadWrite, fsiz - aligned_tail_offset, 0, aligned_tail_offset);
+    }
+    catch (...) {
+        return ignore_checksum;
+    }
+
+    char* tail = map + ((fsiz - tail_size) - aligned_tail_offset);
+
+    // First sum the *last* 128 bytes because if a non-Realm process is overwriting the .realm file
+    // with another valid .realm file it probably does from from the beginning of the file. We want
+    // to catch the end of the original file and the beginning of the new .realm file to get a
+    // checksum mismatch
+
+    for (size_t t = 0; t < tail_size; t++)
+        s += ((tail[t] + 1) * t);
+
+    get_file().unmap(map, fsiz - aligned_tail_offset);
+
+    try {
+        // map() can fail if non-Realm process shortens the file size between get_size() and map()
+        map = (char*)get_file().map(File::access_ReadWrite, head_size);
+    }
+    catch (...) {
+        return ignore_checksum;
+    }
+
+    for (size_t t = 0; t < checksum_offset; t++)
+        s += ((map[t] + 1) * t);
+
+    // Skip the checksum field itself of the header when summing bytes
+    for (size_t t = checksum_offset + 1; t < head_size; t++)
+        s += ((map[t] + 1) * t);
+
+    // Include file size in the checksum
+    s += (fsiz + (fsiz >> 8) + (fsiz >> 16) + (fsiz >> 24));
+
+    get_file().unmap(map, head_size);
+
+    return s;
+}
+
+void SlabAlloc::update_checksum()
+{
+    if (!get_file().is_attached())
+        return;
+
+    char* map = (char*)get_file().map(File::access_ReadWrite, checksum_offset + 1);
+    char c = compute_checksum();
+    reinterpret_cast<Header*>(map)->m_checksum = c;
+    get_file().unmap(map, checksum_offset + 1);
+}
+
+void SlabAlloc::invalidate_checksum()
+{
+    if (!get_file().is_attached())
+        return;
+
+    char* p = (char*)get_file().map(File::access_ReadWrite, checksum_offset + 1);
+    reinterpret_cast<Header*>(p)->m_checksum = ignore_checksum;
+    get_file().unmap(p, checksum_offset + 1);
+}
+
+bool SlabAlloc::verify_checksum()
+{
+    if (!get_file().is_attached())
+        return true;
+
+    char c2 = compute_checksum();
+    char* map;
+
+    try {
+        // map() can fail if non-Realm process shortens the file size between get_size() and map(). Since a
+        // valid .realm file, we return false which means checksum error
+        map = (char*)get_file().map(File::access_ReadWrite, checksum_offset + 1);
+    }
+    catch (...) {
+        return false;
+    }
+
+    char c1 = reinterpret_cast<Header*>(map)->m_checksum;
+
```

Comment:
I think we should have a description of why/how this part of it works. A run-through of the different cases for c1, c2 etc